### PR TITLE
Add bytestring deprecation methods for 0.5

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

**Comment by quinnj** [comment]
Ref: https://github.com/JuliaLang/julia/pull/16453

**Comment by quinnj** [comment]
Hmmm.....how to handle the `Cstring` issue for 0.3? `Cstring` compat isn't defined until line ~698. Do we just have ot move that `bytestring` lower?

**Comment by yuyichao** [comment]
Or move `Cstring` higher. You probably also need a rewrite rule in order to call the type on 0.3

**Comment by quinnj** [comment]
Hmm.....now I'm running into
```julia
julia> reload("Compat")
Warning: replacing module Compat
ERROR: cannot redefine non-generic function
```
I guess we need to handle `bytestring` totally separately on 0.3

**Comment by quinnj** [comment]
Hmm.......I'm struggling with how to get 0.3 compat here. We can't do `typealias String Base.ByteString` because we can't do `call` overloading in 0.3, but if we do `typealias String Union(ASCIIString,UTF8String)`, then we can't define the appropriate constructor methods.

**Comment by yuyichao** [comment]
That's the rewrite rule I'm talking about. You need to use `@compat` to transform a call to `String` to appropriate (helper?) functions on 0.3.

**Comment by quinnj** [comment]
Got it, let's see if I finally got this right.

**Comment by quinnj** [comment]
Anything else we're missing here?

File: src/Compat.jl [hunk 1]
```diff
@@ -4,12 +4,36 @@ module Compat
...
+    export Cstring, Cwstring
+end
+
 if isdefined(Core, :String) && isdefined(Core, :AbstractString)
     # Not exported in order to not break code on 0.5
     typealias UTF8String Core.String
     typealias ASCIIString Core.String
-else
+elseif v"0.4.0" <= VERSION < v"0.5.0"
     typealias String Base.ByteString
```
Comment (yuyichao) [inline src/Compat.jl]: `String` should still be defined on 0.3.

File: src/Compat.jl [hunk 2]
```diff
@@ -4,12 +4,36 @@ module Compat
...
+    call(::Type{Base.ByteString},s::Cstring) = bytestring(s::Cstring)
```
Comment (yuyichao) [inline src/Compat.jl]: Minor issues.

The assertion on the rhs is a little weird. Simple outer constructor syntax should also work.

File: src/Compat.jl (hunk 1 above)
Comment (yuyichao) [inline src/Compat.jl]: And we can test that with `String == @compat(Union{Compat.UTF8String,Compat.ASCIIString})`.

File: src/Compat.jl [hunk 3]
```diff
@@ -485,6 +499,8 @@ function _compat(ex::Expr)
             end
         elseif VERSION < v"0.4.0-dev+4389" && f === :withenv
             rewrite_pairs_to_tuples!(ex)
+        elseif VERSION < v"0.4.0" && f == :String
```
Comment (yuyichao) [inline src/Compat.jl]: It's unfortunate that this is potentially breaking on 0.3 ...

Comment (yuyichao) [inline src/Compat.jl]: Or does `bytestring` covers all usage of `String` on 0.3?

File: src/Compat.jl (hunk 1 above)
Comment (quinnj) [inline src/Compat.jl]: Are you saying something needs to be changed here? Or just a test added?

File: src/Compat.jl (hunk 3 above)
Comment (quinnj) [inline src/Compat.jl]: Hmmm....good question. I think in terms of function calls, it does, but I'm not 100% on that.

File: src/Compat.jl (hunk 1 above)
Comment (yuyichao) [inline src/Compat.jl]: the `typealias` shouldn't be conditional.

File: src/Compat.jl (hunk 3 above)
Comment (yuyichao) [inline src/Compat.jl]: Function call is all what matters here so it should be good then.

File: src/Compat.jl (hunk 1 above)
Comment (quinnj) [inline src/Compat.jl]: Just pushed the change.

File: src/Compat.jl [hunk 4]
```diff
@@ -26,14 +26,16 @@ if isdefined(Core, :String) && isdefined(Core, :AbstractString)
     # Not exported in order to not break code on 0.5
     typealias UTF8String Core.String
     typealias ASCIIString Core.String
-elseif v"0.4.0" <= VERSION < v"0.5.0"
+elseif VERSION < v"0.5.0"
```
Comment (yuyichao) [inline src/Compat.jl]: `else` is enough.

File: src/Compat.jl [hunk 5]
```diff
@@ -4,12 +4,38 @@ module Compat
...
+end
+
 if isdefined(Core, :String) && isdefined(Core, :AbstractString)
     # Not exported in order to not break code on 0.5
     typealias UTF8String Core.String
     typealias ASCIIString Core.String
-else
+elseif VERSION < v"0.5.0"
     typealias String Base.ByteString
+    if VERSION >= v"0.4.0"
```
Comment (tkelman) [inline src/Compat.jl]: please make all VERSION cutoffs as specific as possible

Comment (yuyichao) [inline src/Compat.jl]: This is call overload ecf5426a07bbdd2ab000c9ddc3a461e1cf284027 . Version `0.4.0-dev+1246`

File: src/Compat.jl (hunk 3 above)
Comment (tkelman) [inline src/Compat.jl]: the version cutoff should be the latest version where `String` doesn't cover all signatures of `bytestring`

File: src/Compat.jl [hunk 6]
```diff
@@ -4,12 +4,38 @@ module Compat
...
+    const Cwstring = Ptr{Cwchar_t}
+    export Cstring, Cwstring
+end
+
 if isdefined(Core, :String) && isdefined(Core, :AbstractString)
     # Not exported in order to not break code on 0.5
     typealias UTF8String Core.String
     typealias ASCIIString Core.String
-else
+elseif VERSION < v"0.5.0"
```
Comment (tkelman) [inline src/Compat.jl]: is this needed?

File: src/Compat.jl (hunk 3 above)
Comment (yuyichao) [inline src/Compat.jl]: This is only needed before call overload. Otherwise the definition above should be enough.

File: src/Compat.jl (hunk 6 above)
Comment (yuyichao) [inline src/Compat.jl]: I don't think so .... https://github.com/JuliaLang/Compat.jl/pull/203#discussion_r64139174

**Comment by yuyichao** [comment]
Ping I think this should be good to go after a rebase with the two loose version check fixed and one unnecessary version check removed?

**Comment by quinnj** [comment]
Weird. I did a rebase and force push, but it's not showing up here.....

**Comment by quinnj** [comment]
There should be a single commit now: b6edf4b146adb79d1a87f2532e42d0bf5c31fd10

**Comment by yuyichao** [comment]
Maybe wait longer or try again? It takes some time for your comments to appear on the web interface too....

**Pull request closed without merging**